tokio/src/runtime/task/list.rs

```diff
@@ -1,33 +1,154 @@
+//! This module has containers for storing the tasks spawned on a scheduler. The
+//! `OwnedTasks` container is thread-safe but can only store tasks that
+//! implement Send. The `LocalOwnedTasks` container is not thread safe, but can
+//! store non-Send tasks.
+//!
+//! The collections can be closed to prevent adding new tasks during shutdown of
+//! the scheduler with the collection.
+
+use crate::future::Future;
 use crate::loom::sync::Mutex;
-use crate::runtime::task::Task;
+use crate::runtime::task::{JoinHandle, Notified, RawTask, Schedule, Task};
 use crate::util::linked_list::{Link, LinkedList};
 
+use std::marker::PhantomData;
+
 pub(crate) struct OwnedTasks<S: 'static> {
-    list: Mutex<LinkedList<Task<S>, <Task<S> as Link>::Target>>,
+    inner: Mutex<OwnedTasksInner<S>>,
+}
+struct OwnedTasksInner<S: 'static> {
+    list: LinkedList<Task<S>, <Task<S> as Link>::Target>,
+    closed: bool,
+}
+
+pub(crate) struct LocalOwnedTasks<S: 'static> {
+    list: LinkedList<Task<S>, <Task<S> as Link>::Target>,
+    closed: bool,
+    _not_send: PhantomData<*const ()>,
 }
 
 impl<S: 'static> OwnedTasks<S> {
     pub(crate) fn new() -> Self {
         Self {
-            list: Mutex::new(LinkedList::new()),
+            inner: Mutex::new(OwnedTasksInner {
+                list: LinkedList::new(),
+                closed: false,
+            }),
         }
     }
 
-    pub(crate) fn push_front(&self, task: Task<S>) {
-        self.list.lock().push_front(task);
+    /// Bind the provided task to this OwnedTasks instance. This fails if the
+    /// OwnedTasks has been closed.
+    pub(crate) fn bind<T>(
+        &self,
+        task: T,
+        scheduler: S,
+    ) -> (JoinHandle<T::Output>, Option<Notified<S>>)
+    where
+        S: Schedule,
+        T: Future + Send + 'static,
+        T::Output: Send + 'static,
+    {
+        let raw = RawTask::new::<T, S>(task, scheduler);
```

Comment:
Looks like there is a bunch of duplication between the two `bind` fns (and `task::joinable`). If possible, it would be nice to unify some.